jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/CrossOriginHandler.java

```diff
@@ -0,0 +1,524 @@
+        }
+        else
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("cross-origin request does not match allowed origins: {} {}", request, getAllowedOriginPatterns());
+
+            if (isDeliverNonAllowedOriginRequests())
+            {
+                if (preflight)
+                {
+                    if (!isDeliverPreflightRequests())
+                    {
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("preflight cross-origin request not delivered to child handler {}", request);
+                        callback.succeeded();
+                        return true;
+                    }
+                }
+                else
+                {
+                    if (isWebSocketUpgrade(request))
+                    {
+                        if (!isDeliverNonAllowedOriginWebSocketUpgradeRequests())
+                        {
+                            Response.writeError(request, response, callback, HttpStatus.BAD_REQUEST_400, "origin not allowed");
+                            return true;
+                        }
+                    }
+                }
+
+                if (LOG.isDebugEnabled())
+                    LOG.debug("cross-origin request delivered to child handler {}", request);
+
+                return super.handle(request, response, callback);
+            }
+            else
+            {
+                Response.writeError(request, response, callback, HttpStatus.BAD_REQUEST_400, "origin not allowed");
+                return true;
+            }
+        }
+    }
+
+    private boolean originMatches(String origins)
+    {
+        if (anyOriginAllowed)
+            return true;
+        if (allowedOriginPatterns.isEmpty())
+            return false;
+        return originMatches(origins, allowedOriginPatterns);
+    }
+
+    private boolean timingOriginMatches(String origins)
+    {
+        if (anyTimingOriginAllowed)
+            return true;
+        if (allowedTimingOriginPatterns.isEmpty())
+            return false;
+        return originMatches(origins, allowedTimingOriginPatterns);
+    }
+
+    private boolean originMatches(String origins, Set<Pattern> allowedOriginPatterns)
+    {
+        for (String origin : origins.split(" "))
+        {
+            origin = origin.trim();
+            if (origin.isEmpty())
+                continue;
+            for (Pattern pattern : allowedOriginPatterns)
+            {
+                if (pattern.matcher(origin).matches())
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean isPreflight(Request request)
+    {
+        return HttpMethod.OPTIONS.is(request.getMethod()) && request.getHeaders().contains(HttpHeader.ACCESS_CONTROL_REQUEST_METHOD);
+    }
+
+    private boolean isWebSocketUpgrade(Request request)
+    {
+        return request.getHeaders().contains(HttpHeader.UPGRADE, "websocket");
```

Comment:
I think this will only work for HTTP/1.1?

Reply:
Good point.